_posts/2018-08-30-on-the-fly-transformations.md

```diff
@@ -0,0 +1,283 @@
+Here's what the workflow looks like:
+
+
+```python
+u = mda.Universe('pept_in_memb.tpr', 'pept_in_memb.xtc')
+prot = u.select_atoms("protein")
+# we load another universe to define the reference
+# it uses the same input files, but this doesn't have to be always the case
+ref_u = mda.Universe('pept_in_memb.tpr', 'pept_in_memb.xtc')
+reference = ref_u.select_atoms("protein")
+ag = u.atoms
+transformations = (mda.transformations.unwrap(ag),
+                   mda.transformations.center_in_box(prot, weights='mass'),
+                   mda.transformations.wrap(ag, compound='fragments'),
+                   mda.transformations.fit_rot_trans(prot, reference))
+u.trajectory.add_transformations(*transformations)
+nv.show_mdanalysis(u)
+
+```
+
+
+    NGLWidget(count=11)
+
+
+It looks a bit confusing with the membrane...
+
+
+```python
+t = nv.MDAnalysisTrajectory(prot)
+w = nv.NGLWidget(t)
+w.add_line()
+w
+```
+
+
+    NGLWidget(count=11)
+
+
+This transformations is good when we want to see how the conformation of the protein evolves with time. 
+
+But, in this case, we also have a membrane. How does the protein behave in the membrane? Doing a least squares fitting in the `xy` plane can help us have a better look. Here's how it goes:
+
+
+```python
+u = mda.Universe('pept_in_memb.tpr', 'pept_in_memb.xtc')
+prot = u.select_atoms("protein")
+ref_u = mda.Universe('pept_in_memb.tpr', 'pept_in_memb.xtc')
+reference = ref_u.select_atoms("protein")
+ag = u.atoms
+transformations = (mda.transformations.unwrap(ag),
+                   mda.transformations.center_in_box(prot),
+                   mda.transformations.wrap(ag, compound='fragments'),
+                   mda.transformations.fit_rot_trans(prot, reference, plane='xy', weights="mass"))
+u.trajectory.add_transformations(*transformations)
+# let's hide the lipid tails to have a better view
+view_selection = u.select_atoms("protein or name P")
+t = nv.MDAnalysisTrajectory(view_selection)
+w = nv.NGLWidget(t)
+w.add_line()
+w
+```
+
+
+    NGLWidget(count=11)
+
+
+This transformation keeps the membrane horizontal, while the protein rotation in the z-axis is removed. This becomes particularly useful when observing protein insertion.
+
+### Example 4: I want to do my own transformations...
+The beauty of MDAnalysis transformations is the ability to easily create custom transformations. All transformations must have the following structure:
+    
+```Python
+def custom_transform(args): # arguments at this point are not mandatory
+    #do some things
+        
+    def wrapped(ts): 
+        # this wrapped function must only take a Timestep as argument
+        # this is where the actual changes to the timestep must be done
+            
+        return ts
+        
+    return wrapped
+```
+       
+Let's create one here:
+
+
+```python
+def up_by_2():
+    
+    def wrapped(ts):
+        # here's where the magic happens 
```

Comment:
There is no need for a wrapped function here. The transformation could be written:

```python
def up_by_2(ts):
    ts.positions += np.asarray([0,0,20])
    return ts
```

Reply:
True, but maybe it's easiest to always define transformations in this form so it's not weird when you need to do things with arguments.  So maybe this blogpost shouldn't have this simple case that can flatten out @davidercruz

Reply:
In an ideal world, this would be the perfect example. However, I would just add a footnote saying that in this particular case we could also write as https://github.com/MDAnalysis/MDAnalysis.github.io/pull/87/files#r209500911 says but the pattern described here is general and will also work for situations when the simple approach doesn't work.